kube-derive/src/custom_resource.rs

```diff
@@ -375,15 +364,93 @@ pub(crate) fn derive(input: proc_macro2::TokenStream) -> proc_macro2::TokenStrea
             fn api_resource() -> kube::core::ApiResource {
                 kube::core::ApiResource::erase::<Self>(&())
             }
+
         }
     };
 
+    let impl_hasspec = generate_hasspec(&ident, &rootident);
+
     // Concat output
     quote! {
         #root_obj
         #impl_resource
         #impl_default
         #impl_crd
+        #impl_hasspec
+        #impl_hasstatus
+    }
+}
+
+
+/// This generates the code for the `kube::core::object::HasSpec` trait implementation.
+///
+/// All CRDs have a spec so it is implemented for all of them.
+///
+/// # Arguments
+///
+/// * `ident`: The identity (name) of the spec struct
+/// * `root ident`: The identity (name) of the main CRD struct (the one we generate in this macro)
+fn generate_hasspec(spec_ident: &Ident, root_ident: &Ident) -> TokenStream {
+    quote! {
+        impl kube::core::object::HasSpec for #root_ident {
+            type Spec = #spec_ident;
+
+            fn spec(&self) -> &#spec_ident {
+                &self.spec
+            }
+
+            fn spec_mut(&mut self) -> &mut #spec_ident {
+                &mut self.spec
+            }
+        }
+    }
+}
+
+
+/// This processes the `status` field of a CRD.
+///
+/// As it is optional some features will be turned on or off depending on whether it's available or not.
+///
+/// # Arguments
+///
+/// * `root ident`: The identity (name) of the main CRD struct (the one we generate in this macro)
+/// * `status`: The optional name of the `status` struct to use
+/// * `visibility`: Desired visibility of the generated field
+///
+/// returns: A tuple of (status_field, status_default, status_impl)
+/// * `status_field` is the code to be used for the field in the main struct
+/// * `status_default` is the initialization code to use in a `Default` implementation
+/// * `status_impl` is the implementation code for the `HasStatus` trait
+fn process_status(root_ident: &Ident, status: &Option<String>, visibility: &Visibility) -> (TokenStream, TokenStream, TokenStream) {
+    if let Some(status_name) = &status {
+        let ident = format_ident!("{}", status_name);
+
+        let status_field = quote! {
+            #[serde(skip_serializing_if = "Option::is_none")]
+            #visibility status: Option<#ident>,
+        };
+        let status_default = quote! { status: None, };
+
+        let status_impl = quote! {
+            impl kube::core::object::HasStatus for #root_ident {
+
+                type Status = #ident;
+
+                fn status(&self) -> Option<&#ident> {
+                    self.status.as_ref()
+                }
+
+                fn status_mut(&mut self) -> &mut Option<#ident> {
+                    &mut self.status
+                }
+            }
+        };
+        (status_field, status_default, status_impl)
```

Comment:
Nit: This is almost getting to the point where we might want to consider moving it into a struct with named fields (especially considering that they all have the same type).

Not really a blocker for the PR though.

Reply:
Happy to do that though.. I'll get on it. The hardest part will be the naming ;-)